tests/regression/src/lib.rs

```diff
@@ -0,0 +1,143 @@
+        }
+        run_valgrind_test("test_rsa_handshake");
+        Ok(())
+    }
+    // function to run specified test using valgrind
+    fn run_valgrind_test(test_name: &str) {
+        let exe_path = std::env::args().next().unwrap();
+        let output_file = format!("cachegrind_{}.out", test_name);
+        let valgrind_command = format!(
+            "valgrind --tool=cachegrind --cachegrind-out-file={} {} {}",
+            output_file, exe_path, test_name
+        );
+    
+        println!("Running command: {}", valgrind_command);
+        let output_command = format!("--cachegrind-out-file={}", &output_file);
+        let status = Command::new("valgrind")
+            .args(&["--tool=cachegrind", &output_command, &exe_path, test_name])
+            .env_remove("VALGRIND") //ensures that the recursive call is made to the actual harness code block rather than back to this function
+            .status()
+            .expect("Failed to execute cargo test");
+    
+        if !status.success() {
+            panic!("Valgrind failed");
+        }
+    
+        let annotate_file = format!("perf_outputs/{}.annotated.txt", test_name);
+        // this command annotates the raw output file and pipes it to the specified location
+        let annotate_command = format!(
+            "cg_annotate {} > {}",
+            output_file, annotate_file
+        );
+    
+        println!("Running command: {}", annotate_command);
+    
+        let annotate_status = Command::new("sh")
+            .arg("-c")
+            .arg(&annotate_command)
+            .status()
+            .expect("Failed to run cg_annotate");
+
+        if !annotate_status.success() {
+            panic!("cg_annotate failed");
+        }
+    
+        let count = grep_for_instructions(&annotate_file).expect("Failed to get instruction count from file");
+        //this is temporary code to showcase the future diff functionality, here the code regresses by 10% each time so this test will almost always fail in its current state
+        let new_count = count + count / 10;
+        let diff = new_count - count;
+        assert!(diff <= self::COST, "Instruction count difference in {} exceeds the threshold, regression of {} instructions", test_name, diff);
+    }
+
+    use std::fs::File;
+    use std::io::{self, BufRead};
+    use std::path::Path;
+    // parses the annotated file for the overall instruction count total
+    fn grep_for_instructions(file_path: &str) -> Result<u64, io::Error> {
+        let path = Path::new(file_path);
+        let file = File::open(&path)?;
+        let reader = io::BufReader::new(file);
+
+        for line in reader.lines() {
+            let line = line?;
+            if line.contains("PROGRAM TOTALS") {
+                let parts: Vec<&str> = line.split_whitespace().collect();
+                if let Some(instructions) = parts.get(0) {
+                    if let Ok(instructions) = instructions.replace(",", "").parse::<u64>() {
```

Comment:
did you run `clippy` on this? I think it would be complaining about this
```suggestion
                    if let Ok(instructions) = instructions.replace(',', "").parse::<u64>() {
```

Reply:
It didn't complain but I cleaned it up a bit in the new revision